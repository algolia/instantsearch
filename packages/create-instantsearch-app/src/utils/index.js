const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

const algoliasearch = require('algoliasearch');
const chalk = require('chalk');
const semver = require('semver');
const validateProjectName = require('validate-npm-package-name');

const TEMPLATES_FOLDER = path.join(__dirname, '../templates');

const algoliaConfig = {
  appId: 'OFCNCOG2CU',
  apiKey: 'f54e21fa3a2a0160595bb058179bfb1e',
  indexName: 'npm-search',
};

const client = algoliasearch(algoliaConfig.appId, algoliaConfig.apiKey);
const index = client.initIndex(algoliaConfig.indexName);

function checkAppName(appName) {
  const validationResult = validateProjectName(appName);

  if (!validationResult.validForNewPackages) {
    let errorMessage = `Could not create a project called "${chalk.red(
      appName
    )}" because of npm naming restrictions.`;

    (validationResult.errors || []).forEach((error) => {
      errorMessage += `\n  - ${error}`;
    });

    throw new Error(errorMessage);
  }

  return true;
}

function checkAppPath(appPath) {
  if (fs.existsSync(appPath)) {
    if (fs.lstatSync(appPath).isDirectory()) {
      const files = fs.readdirSync(appPath);

      if (files && files.length > 0) {
        throw new Error(
          `Could not create project in destination folder "${chalk.red(
            appPath
          )}" because it is not empty.`
        );
      }
    } else {
      throw new Error(
        `Could not create project at path ${chalk.red(
          appPath
        )} because a file of the same name already exists.`
      );
    }
  }

  if (!appPath) {
    throw new Error('Could not create project without directory');
  }

  return true;
}

function getAppTemplateConfig(templatePath, { loadFileFn = require } = {}) {
  try {
    return loadFileFn(path.join(templatePath, '.template.js'));
  } catch (err) {
    throw new Error(
      `The template configuration file \`.template.js\` contains errors:
${err.message}`
    );
  }
}

function isYarnAvailable() {
  try {
    execSync('yarnpkg --version', { stdio: 'ignore' });
    return true;
  } catch (err) {
    return false;
  }
}

function getAllTemplates() {
  const templates = fs
    .readdirSync(TEMPLATES_FOLDER)
    .map((name) => path.join(TEMPLATES_FOLDER, name))
    .filter((source) => fs.lstatSync(source).isDirectory())
    .map((source) => path.basename(source));

  return templates;
}

function getTemplatesByCategory() {
  const templatePaths = fs
    .readdirSync(TEMPLATES_FOLDER)
    .map((name) => path.join(TEMPLATES_FOLDER, name))
    .filter((source) => fs.lstatSync(source).isDirectory());

  const templates = templatePaths.reduce((allTemplates, source) => {
    const name = path.basename(source);

    const { category } = require(`${source}/.template.js`);

    if (!category) {
      return allTemplates;
    }

    return {
      ...allTemplates,
      [category]: [...(allTemplates[category] || []), name],
    };
  }, {});

  return templates;
}

function getTemplatePath(templateName) {
  const supportedTemplates = getAllTemplates();

  // We support the template, let's retrieve its path
  if (supportedTemplates.includes(templateName)) {
    return path.join(TEMPLATES_FOLDER, templateName);
  }

  // This is a custom template, it's a path already
  return templateName;
}

async function fetchLibraryVersions(libraryName) {
  const library = await index.getObject(libraryName, {
    attributesToRetrieve: ['versions'],
  });

  return Object.keys(library.versions).reverse();
}

function getLibraryVersion({ libraryName, supportedVersion = '' }) {
  return async (getVersion) => {
    const versions = await fetchLibraryVersions(libraryName);

    return getVersion(versions, supportedVersion);
  };
}

async function getLatestLibraryVersion(...args) {
  return await getLibraryVersion(...args)(semver.maxSatisfying);
}

async function getEarliestLibraryVersion(...args) {
  return await getLibraryVersion(...args)(semver.minSatisfying);
}

const splitArray = (string) =>
  string
    .split(',')
    .filter(Boolean)
    .map((x) => x.trim());

module.exports = {
  checkAppName,
  checkAppPath,
  getAppTemplateConfig,
  isYarnAvailable,
  fetchLibraryVersions,
  getLatestLibraryVersion,
  getEarliestLibraryVersion,
  getAllTemplates,
  getTemplatePath,
  getTemplatesByCategory,
  splitArray,
};
