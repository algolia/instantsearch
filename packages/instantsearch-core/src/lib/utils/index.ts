// This folder is used for utility functions that are useful within the InstantSearch Core library. Not for external usage. For that the `src/lib/public` folder should be used.

export * from './addAbsolutePosition';
export * from './addQueryID';
export * from './checkIndexUiState';
export * from './checkRendering';
export * from './clearRefinements';
export * from './createSendEventForFacet';
export * from './createSendEventForHits';
export * from './escapeFacetValue';
export * from './geo-search';
export * from './getAppIdAndApiKey';
export * from './getInsightsAnonymousToken';
export * from './getRefinements';
export * from './hydrateRecommendCache';
export * from './hydrateSearchClient';
export * from './mergeSearchParameters';
export * from './resolveSearchParameters';
export * from './setIndexHelperState';
