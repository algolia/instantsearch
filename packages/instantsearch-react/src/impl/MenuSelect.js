import createFacetRefiner from '../createFacetRefiner';

import Select from './Select';

export default createFacetRefiner(Select);
