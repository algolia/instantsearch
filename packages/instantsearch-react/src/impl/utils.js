export function isSpecialClick(event) {
  const isMiddleClick = event.button === 1;
  return Boolean(
    isMiddleClick ||
    event.altKey ||
    event.ctrlKey ||
    event.metaKey ||
    event.shiftKey
  );
}

export function getTranslation(key, defaults, custom, ...params) {
  const translation =
    {}.hasOwnProperty.call(custom, key) ?
      custom[key] :
      defaults[key];
  if (typeof translation === 'function') {
    return translation(...params);
  }
  return translation;
}

export function capitalize(key) {
  return key.length === 0 ? '' : `${key[0].toUpperCase()}${key.slice(1)}`;
}
