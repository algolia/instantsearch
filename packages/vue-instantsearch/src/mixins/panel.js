import mitt from 'mitt';

import { isVue3 } from '../util/vue-compat';

export const PANEL_EMITTER_NAMESPACE = 'instantSearchPanelEmitter';
export const PANEL_CHANGE_EVENT = 'PANEL_CHANGE_EVENT';

export const createPanelProviderMixin = () => ({
  props: {
    emitter: {
      type: Object,
      required: false,
      default() {
        return mitt();
      },
    },
  },
  provide() {
    return {
      [PANEL_EMITTER_NAMESPACE]: this.emitter,
    };
  },
  data() {
    return {
      canRefine: true,
    };
  },
  created() {
    this.emitter.on(PANEL_CHANGE_EVENT, (value) => {
      this.updateCanRefine(value);
    });
  },
  [isVue3 ? 'beforeUnmount' : 'beforeDestroy']() {
    this.emitter.all.clear();
  },
  methods: {
    updateCanRefine(value) {
      this.canRefine = value;
    },
  },
});

export const createPanelConsumerMixin = ({
  mapStateToCanRefine = (state) => Boolean(state.canRefine),
} = {}) => ({
  inject: {
    emitter: {
      from: PANEL_EMITTER_NAMESPACE,
      default() {
        return {
          emit: () => {},
        };
      },
    },
  },
  data() {
    return {
      state: null,
      hasAlreadyEmitted: false,
    };
  },
  watch: {
    state: {
      immediate: true,
      handler(nextState, previousState) {
        if (!nextState) {
          return;
        }

        const previousCanRefine = mapStateToCanRefine(previousState || {});
        const nextCanRefine = mapStateToCanRefine(nextState);

        if (!this.hasAlreadyEmitted || previousCanRefine !== nextCanRefine) {
          this.emitter.emit(PANEL_CHANGE_EVENT, nextCanRefine);
          this.hasAlreadyEmitted = true;
        }
      },
    },
  },
});
