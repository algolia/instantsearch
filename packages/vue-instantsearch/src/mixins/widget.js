import { isVue3 } from '../util/vue-compat';
import { warn } from '../util/warn';

export const createWidgetMixin = (
  { connector } = {},
  additionalProperties = {}
) => ({
  inject: {
    instantSearchInstance: {
      from: '$_ais_instantSearchInstance',
      default() {
        const tag = this.$options._componentTag;
        throw new TypeError(
          `It looks like you forgot to wrap your Algolia search component "<${tag}>" inside of an "<ais-instant-search>" component.`
        );
      },
    },
    getParentIndex: {
      from: '$_ais_getParentIndex',
      default() {
        return () => this.instantSearchInstance.mainIndex;
      },
    },
  },
  data() {
    return {
      state: null,
    };
  },
  created() {
    if (typeof connector === 'function') {
      this.factory = connector(this.updateState, () => {});
      this.widget = {
        ...this.factory(this.widgetParams),
        ...additionalProperties,
      };
      this.getParentIndex().addWidgets([this.widget]);

      if (
        this.instantSearchInstance._initialResults &&
        !this.instantSearchInstance.started
      ) {
        if (typeof this.instantSearchInstance.__forceRender !== 'function') {
          throw new Error(
            'You are using server side rendering with <ais-instant-search> instead of <ais-instant-search-ssr>.'
          );
        }
        this.instantSearchInstance.__forceRender(
          this.widget,
          this.getParentIndex()
        );
      }
    } else if (connector !== true) {
      warn(
        `You are using the InstantSearch widget mixin, but didn't provide a connector.
While this is technically possible, and will give you access to the Helper,
it's not the recommended way of making custom components.

If you want to disable this message, pass { connector: true } to the mixin.

Read more on using connectors: https://alg.li/vue-custom`
      );
    }
  },
  [isVue3 ? 'beforeUnmount' : 'beforeDestroy']() {
    if (this.widget) {
      this.getParentIndex().removeWidgets([this.widget]);
    }
  },
  watch: {
    widgetParams: {
      handler(nextWidgetParams) {
        this.state = null;
        this.getParentIndex().removeWidgets([this.widget]);
        this.widget = {
          ...this.factory(nextWidgetParams),
          ...additionalProperties,
        };
        this.getParentIndex().addWidgets([this.widget]);
      },
      deep: true,
    },
  },
  methods: {
    updateState(state = {}, isFirstRender) {
      if (!isFirstRender) {
        // Avoid updating the state on first render
        // otherwise there will be a flash of placeholder data
        this.state = state;
      }
    },
  },
});
