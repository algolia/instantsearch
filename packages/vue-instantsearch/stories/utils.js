import { liteClient as algoliasearch } from 'algoliasearch/lite';

export const previewWrapper =
  ({
    searchClient = algoliasearch('latency', '6be0576ff61c053d5f9a3225e2a90f76'),
    insightsClient,
    indexName = 'instant_search',
    hits = `
    <template v-slot="{ items }">
      <ol class="playground-hits">
        <li
          v-for="item in items"
          :key="item.objectID"
          class="playground-hits-item"
        >
          <div
            class="playground-hits-image"
            :style="{ backgroundImage: 'url(' + item.image + ')' }"
          />
          <div class="playground-hits-desc">
            <p>
              <ais-highlight attribute="name" :hit="item" />
            </p>
            <p>Rating: {{ item.rating }}âœ­</p>
            <p>Price: {{ item.price }}$</p>
          </div>
        </li>
      </ol>
    </template>
  `,
    filters = `
    <ais-refinement-list attribute="brand" />
    <ais-configure :hitsPerPage="3" />
  `,
    routing,
  } = {}) =>
  () => ({
    template: `
    <ais-instant-search
      :search-client="searchClient"
      index-name="${indexName}"
      :routing="routing"
      :insights-client="insightsClient"
    >
      <div class="vis-container vis-container-preview">
        <story />
      </div>

      <div class="vis-container vis-container-playground">
        <div class="panel-left">
          ${filters}
        </div>
        <div class="panel-right">
          <ais-search-box />
          <ais-stats />
          <ais-hits>
            ${hits}
          </ais-hits>
          <ais-pagination />
        </div>
      </div>
    </ais-instant-search>
  `,
    data() {
      return {
        searchClient,
        routing,
        insightsClient,
      };
    },
  });
