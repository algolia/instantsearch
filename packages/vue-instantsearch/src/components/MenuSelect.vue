<template>
  <div
    :class="[suit(), !state.canRefine && suit('', 'noRefinement')]"
    v-if="state"
  >
    <slot
      :items="state.items"
      :can-refine="state.canRefine"
      :refine="refine"
      :createURL="state.createURL"
      :send-event="state.sendEvent"
    >
      <select
        :class="suit('select')"
        @change="refine($event.currentTarget.value)"
      >
        <option :class="suit('option')" value="">
          <slot name="defaultOption"> See all </slot>
        </option>
        <option
          v-for="item in state.items"
          :key="item.value"
          :class="suit('option')"
          :value="item.value"
          :selected="item.isRefined"
        >
          <slot name="item" :item="item">
            {{ item.label }} ({{ item.count }})
          </slot>
        </option>
      </select>
    </slot>
  </div>
</template>

<script>
import { connectMenu } from 'instantsearch.js/es/connectors';

import { createPanelConsumerMixin } from '../mixins/panel';
import { createSuitMixin } from '../mixins/suit';
import { createWidgetMixin } from '../mixins/widget';

export default {
  name: 'AisMenuSelect',
  mixins: [
    createSuitMixin({ name: 'MenuSelect' }),
    createWidgetMixin(
      { connector: connectMenu },
      {
        $$widgetType: 'ais.menuSelect',
      }
    ),
    createPanelConsumerMixin(),
  ],
  props: {
    attribute: {
      type: String,
      required: true,
    },
    limit: {
      type: Number,
      default: 10,
    },
    sortBy: {
      type: [Array, Function],
      default: undefined,
    },
    transformItems: {
      type: Function,
      default(items) {
        return items;
      },
    },
  },
  computed: {
    widgetParams() {
      return {
        attribute: this.attribute,
        limit: this.limit,
        sortBy: this.sortBy,
        transformItems: this.transformItems,
      };
    },
  },
  methods: {
    refine(value) {
      this.state.refine(value);
    },
  },
};
</script>
