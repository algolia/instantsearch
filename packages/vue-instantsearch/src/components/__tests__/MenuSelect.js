/**
 * @jest-environment jsdom
 */

import { mount } from '../../../test/utils';
import { __setState } from '../../mixins/widget';
import MenuSelect from '../MenuSelect.vue';
import '../../../test/utils/sortedHtmlSerializer';

jest.mock('../../mixins/widget');
jest.mock('../../mixins/panel');

const defaultState = {
  canRefine: true,
  items: [
    { label: 'Apple', value: 'Apple', isRefined: false, count: 50 },
    { label: 'Samsung', value: 'Samsung', isRefined: false, count: 20 },
    { label: 'Sony', value: 'Sony', isRefined: false, count: 15 },
  ],
};

const defaultProps = {
  attribute: 'brand',
};

describe('default render', () => {
  it('renders correctly with custom label', () => {
    __setState({
      ...defaultState,
    });

    const props = {
      ...defaultProps,
    };

    const wrapper = mount(MenuSelect, {
      propsData: props,
      slots: {
        // tag is needed here for Vue Test Utils, even if it's invalid HTML
        defaultOption: '<span>None</span>',
      },
    });

    expect(wrapper.find('option').html()).toContain('None');
    expect(wrapper.html()).toMatchSnapshot();
  });
});

it('exposes send-event method for insights middleware', async () => {
  const sendEvent = jest.fn();
  __setState({
    ...defaultState,
    sendEvent,
  });

  const wrapper = mount({
    components: { MenuSelect },
    data() {
      return { props: defaultProps };
    },
    template: `
      <MenuSelect v-bind="props">
        <template v-slot="{ sendEvent }">
          <div>
            <button @click="sendEvent()">Send Event</button>
          </div>
        </template>
      </MenuSelect>
    `,
  });

  await wrapper.find('button').trigger('click');
  expect(sendEvent).toHaveBeenCalledTimes(1);
});

describe('custom item slot', () => {
  it('renders correctly', () => {
    __setState({
      ...defaultState,
    });

    const wrapper = mount({
      components: { MenuSelect },
      data() {
        return { props: defaultProps };
      },
      template: `
        <MenuSelect v-bind="props">
          <template v-slot:item="{ item }">
            <span>
              {{ item.label.toUpperCase() }}
            </span>
          </template>
        </MenuSelect>
      `,
    });

    expect(wrapper.html()).toMatchSnapshot();

    expect(wrapper.find('option:nth-child(2)').html()).toMatch(
      /<span>\s*APPLE\s*<\/span>/
    );
  });
});

describe('custom default render', () => {
  const defaultSlot = `
    <template v-slot="{ items, canRefine, refine }">
      <select
        @change="refine($event.currentTarget.value)"
        :disabled="!canRefine"
      >
        <option value="">All</option>
        <option
          v-for="item in items"
          :key="item.value"
          :value="item.value"
          :selected="item.isRefined"
        >
          {{item.label.toUpperCase()}}
        </option>
      </select>
    </template>
  `;

  it('renders correctly', () => {
    __setState({
      ...defaultState,
    });

    const wrapper = mount({
      components: { MenuSelect },
      data() {
        return { props: defaultProps };
      },
      template: `
        <MenuSelect v-bind="props">
          ${defaultSlot}
        </MenuSelect>
      `,
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders correctly with a selected value', () => {
    __setState({
      ...defaultState,
      items: [
        { label: 'Apple', value: 'Apple', isRefined: false, count: 50 },
        { label: 'Samsung', value: 'Samsung', isRefined: true, count: 20 },
        { label: 'Sony', value: 'Sony', isRefined: false, count: 15 },
      ],
    });

    const wrapper = mount({
      components: { MenuSelect },
      data() {
        return { props: defaultProps };
      },
      template: `
        <MenuSelect v-bind="props">
          <template v-slot="{ items, canRefine, refine }">
            <select
              @change="refine($event.currentTarget.value)"
              :disabled="!canRefine"
            >
              <option value="">All</option>
              <option
                v-for="item in items"
                :key="item.value"
                :value="item.value"
                :selected="item.isRefined"
                :data-test="item.value"
              >
                {{item.label}}
              </option>
            </select>
          </template>
        </MenuSelect>
      `,
    });

    // Unlike in vue 2, in vue 3 when value and textContent are the same,
    // Vue doesn't render the value attribute at all.
    // For example,

    // In Vue 2,
    // <option value="Apple">Apple</option>

    // becomes, in Vue 3
    // <option>Apple</option>

    const selected = wrapper.find('[data-test="Samsung"]');
    const options = wrapper.findAll('option:not([data-test="Samsung"])');

    expect(selected.element.selected).toBe(true);

    for (let i = 0; i < options.length; i++) {
      expect((options[i] || options.at(i)).element.selected).toBe(false);
    }
  });

  it('renders correctly without refinements', () => {
    __setState({
      ...defaultState,
      canRefine: false,
      items: [],
    });

    const wrapper = mount({
      components: { MenuSelect },
      data() {
        return { props: defaultProps };
      },
      template: `
        <MenuSelect v-bind="props">
          ${defaultSlot}
        </MenuSelect>
      `,
    });

    expect(wrapper.htmlCompat()).toMatchSnapshot();
  });

  it('calls refine on select change', async () => {
    const refine = jest.fn();

    __setState({
      ...defaultState,
      refine,
    });

    const wrapper = mount({
      components: { MenuSelect },
      data() {
        return { props: defaultProps };
      },
      template: `
        <MenuSelect v-bind="props">
          ${defaultSlot}
        </MenuSelect>
      `,
    });

    expect(refine).not.toHaveBeenCalled();

    const select = wrapper.find('select');

    // Simulate the change
    select.element.value = 'Apple';

    await select.trigger('change');

    expect(refine).toHaveBeenCalledTimes(1);
    expect(refine).toHaveBeenCalledWith('Apple');
  });
});
