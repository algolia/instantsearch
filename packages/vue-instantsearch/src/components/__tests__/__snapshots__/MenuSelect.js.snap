// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom default render renders correctly 1`] = `
<div class="ais-MenuSelect">
  <select>
    <option value>
      All
    </option>
    <option value="Apple">
      APPLE
    </option>
    <option value="Samsung">
      SAMSUNG
    </option>
    <option value="Sony">
      SONY
    </option>
  </select>
</div>
`;

exports[`custom default render renders correctly without refinements 1`] = `
<div class="ais-MenuSelect ais-MenuSelect--noRefinement">
  <select disabled="disabled">
    <option value>
      All
    </option>
  </select>
</div>
`;

exports[`custom item slot renders correctly 1`] = `
<div class="ais-MenuSelect">
  <select class="ais-MenuSelect-select">
    <option class="ais-MenuSelect-option"
            value
    >
      See all
    </option>
    <option class="ais-MenuSelect-option"
            value="Apple"
    >
      <span>
        APPLE
      </span>
    </option>
    <option class="ais-MenuSelect-option"
            value="Samsung"
    >
      <span>
        SAMSUNG
      </span>
    </option>
    <option class="ais-MenuSelect-option"
            value="Sony"
    >
      <span>
        SONY
      </span>
    </option>
  </select>
</div>
`;

exports[`default render renders correctly with custom label 1`] = `
<div class="ais-MenuSelect">
  <select class="ais-MenuSelect-select">
    <option class="ais-MenuSelect-option"
            value
    >
      <span>
        None
      </span>
    </option>
    <option class="ais-MenuSelect-option"
            value="Apple"
    >
      Apple (50)
    </option>
    <option class="ais-MenuSelect-option"
            value="Samsung"
    >
      Samsung (20)
    </option>
    <option class="ais-MenuSelect-option"
            value="Sony"
    >
      Sony (15)
    </option>
  </select>
</div>
`;
